### DISCUSSION TITLE
How to upgrade Talos and k8s using OpenTofu

### DISCUSSION THREAD

**Discussion by rancherr** [opentofu/opentofu#1882] [General]
Hei!
I am new to opentofu.
I use talos and k8s for driving the prod and I would like to automate the talos and k8s upgrades.

I tried with ansible but its not so good as ansible has no talos provider and its just running the scripts|commands inside which is not the point.
Since I would like to have all opensource tools I can see that opentofuhas talos providers but cant see any upgrade instructions.

Could anyone help me to find out how to do it?

**Comment by Evi1Pumpkin** [comment]
Hey @rancherr,
Welcome to the OpenTofu community! I'm sorry to hear that you're facing some problems.
Unfortunately, I can't help you with that issue, as I've never used Talos. 
If I were in your shoes, I would start by finding out if the Talos provider supports upgrading K8s. You can visit their [repository](https://github.com/hcloud-talos/terraform-hcloud-talos) and open an issue to ask them if they support it and, if they do, how to go about it. Good luck!

**Reply by rancherr** [reply]
Thank you so much for the answer.
First I would like to start with Talos upgrades on many nodes... 
Ansible - no Talos module...
So I wanted to use OpenTofu to take care of my  Talos/k8s infra - with upgrades...

**Reply by ghost** [reply]
@rancherr I think an important distinction to make between Ansible and OpenTofu is that with Ansible you can simply upgrade your setup even if you didn't use Ansible to create it in the first place. OpenTofu manages the full lifecycle of resources, so in order to manage the resources you will need to write a config matching your setup and either create your setup with OpenTofu or import your existing resources.

It is also worth noting that a lot of OpenTofu providers don't implement in-place upgrades and resources (such as VMs, etc.) will be destroyed and replaced by the upgrade process, implementing the "immutable infrastructure" idea.
